unity-renderer/Assets/Rendering/Utils/GPUSkinning/SimpleGPUSkinning.cs

```diff
@@ -21,28 +23,59 @@ public static void EncodeBindPosesIntoMesh(SkinnedMeshRenderer skr)
         {
             Mesh sharedMesh = skr.sharedMesh;
             int vertexCount = sharedMesh.vertexCount;
-            Vector4[] bone01data = new Vector4[vertexCount];
-            Vector4[] bone23data = new Vector4[vertexCount];
-
-            BoneWeight[] boneWeights = sharedMesh.boneWeights;
+            
+            NativeArray<Vector4> bone01data = new NativeArray<Vector4>(vertexCount, Allocator.Temp, NativeArrayOptions.UninitializedMemory);
+            NativeArray<Vector4> bone23data = new NativeArray<Vector4>(vertexCount, Allocator.Temp, NativeArrayOptions.UninitializedMemory);
 
+            var bonesPerVertex = sharedMesh.GetBonesPerVertex();
+            var boneWeightsNative = sharedMesh.GetAllBoneWeights();
+            var boneIndex = 0;
             for ( int i = 0; i < vertexCount; i ++ )
             {
-                BoneWeight boneWeight = boneWeights[i];
-                bone01data[i].x = boneWeight.boneIndex0;
-                bone01data[i].y = boneWeight.weight0;
-                bone01data[i].z = boneWeight.boneIndex1;
-                bone01data[i].w = boneWeight.weight1;
-
-                bone23data[i].x = boneWeight.boneIndex2;
-                bone23data[i].y = boneWeight.weight2;
-                bone23data[i].z = boneWeight.boneIndex3;
-                bone23data[i].w = boneWeight.weight3;
+                byte boneCount = bonesPerVertex[i];
+                if (boneCount == 0)
+                {
+                    bone01data[i] = new Vector4(0, 0, 0, 0);
+                    bone23data[i] = new Vector4(0, 0, 0, 0);
+                } 
+                else if (boneCount == 1)
+                {
+                    var bw0 = boneWeightsNative[boneIndex];
+                    bone01data[i] = new Vector4(bw0.boneIndex, bw0.weight, 0, 0);
```

Comment:
Isn't `new Vector` allocating memory? and the assignment copying it? If you are already initializing the memory with zeros, then something like this may be better:

```suggestion
var vec = bone01data[i];
vec.x = bw0.boneIndex;
vec.y = bw0.weight;
```

It is not the prettiest, but in search of speed some stylish things can be put in second place to reduce operations